pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,437 @@
+package registrysyncer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	crcontrollerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	imagev1 "github.com/openshift/api/image/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	controllerutil "github.com/openshift/ci-tools/pkg/controller/util"
+	"github.com/openshift/ci-tools/pkg/util/imagestreamtagmapper"
+	"github.com/openshift/ci-tools/pkg/util/imagestreamtagwrapper"
+)
+
+const ControllerName = "registry_syncer"
+
+func AddToManager(mgr manager.Manager,
+	managers map[string]manager.Manager,
+	pullSecretGetter func() []byte,
+	dryRun bool,
+) error {
+	log := logrus.WithField("controller", ControllerName)
+	r := &reconciler{
+		ctx:              context.Background(),
+		log:              log,
+		registryClients:  map[string]ctrlruntimeclient.Client{},
+		pullSecretGetter: pullSecretGetter,
+		dryRun:           dryRun,
+	}
+	for ctx, m := range managers {
+		r.registryClients[ctx] = imagestreamtagwrapper.MustNew(m.GetClient(), m.GetCache())
+	}
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler: r,
+		// We conflict on ImageStream level which means multiple request for imagestreamtags
+		// of the same imagestream will conflict so stay at one worker in order to reduce the
+		// number of errors we see. If we hit performance issues, we will probably need cluster
+		// and/or imagestream level locking.
+		MaxConcurrentReconciles: 1,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	allContexts := sets.NewString()
+	for ctx := range managers {
```

Comment:
Nit: You can move this in the loop below, no reason to loop twice

Reply:
Need to calculate `allContexts ` first and use it in the 2nd loop.

Reply:
Right, except that you don't need it there because you don't need to create cluster-scoped `reconcile.Requests`